### PR TITLE
[Example] Heterograph convolution with ArangoDB for incidence management

### PR DESCRIPTION
**Pull request by rajivsam**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by jermainewang** [comment]
Hi @rajivsam , welcome to the DGL community. As an initial step, could you please describe what this PR is about? What does it achieve and how does it compare with the existing RGCN examples in our repo?

Also, the +312,099 lines of changes look pretty scary. Did you accidentally include any data files?

**Comment by BarclayII** [comment]
Hi,

Do you intend to contribute this to our tutorial or example collection?  If so, could you give us some background on the problem you were solving, and the methodology behind it?

Thanks.

**Comment by rajivsam** [comment]
Hi Andy,
Thank you for your response. I thought I was submitting this to an internal
repository. Anyway, it is all good and here is the background. My name is
Rajiv Sambasivan and I work with Machine Learning at ArangoDB. We are
interested in contributing to this to the example collection. The
background for this contribution is as follows. We were exploring DGL. We
were interested in developing embeddings for graphs stored in ArangoDB. To
explore this, we used a dataset from the UCI machine learning repository:
https://archive.ics.uci.edu/ml/datasets/Incident+management+process+enriched+event+log
The data is the event log from incident management (help desk). The data is
a dump from a relational database. We modeled this as a graph problem with
the following abstractions:

   1. The  Incident
   2. The support organization assigned to work on resolving the incident
   3. The customer who reported the incident
   4. The vendor or product associated with the incident

Each of these abstractions forms the vertices of a graph. Each vertex has a
set of features. Since each vertex is a distinct entity, the graph is a
heterogeneous one with features for each node. The task is to develop the
embeddings associated with these abstractions. The analysis consists of the
following steps. We have a notebook corresponding to each step:

   1. Exploratory Data Analysis: We analyze the data, examine the data
   types and pick a node attribute to predict. For this problem, we picked an
   engineered attribute. The attribute to predict captures if the ticket is
   reassigned. The rationale for picking this feature was the following. This
   attribute indicates if the process of the resolution was done effectively.
   Importantly, this attribute has a good balance from the point of view of
   supervised learning. About 50% of the tickets are reassigned and is the
   property/attribute of the incident vertex. The analysis is captured in
   https://github.com/arangoml/dgl/blob/managed_services_version_with_sampling/tutorials/models/hetro-graph-with-node-features/ITSM_EDA.ipynb
   .
   2. Data Preprocessing: This step prepares the data for analysis using
   DGL. The data is a dump from a relational database. This data must be
   massaged to a form that is suitable for analysis by DGL using Pytorch as
   the backend. The activity is captured in the notebook
   https://github.com/arangoml/dgl/blob/managed_services_version_with_sampling/tutorials/models/hetro-graph-with-node-features/batch_graph_pre-processing.ipynb
   3. Data loading utility development: We chose Pytorch as the backend for
   DGL. The data loader loads the preprocessed data from the previous step
   into ArangoDB. DGL works with a NetworkX version of the graph. So the data
   loader also provides this. The data loader is captured in the following
   notebook:
   https://github.com/arangoml/dgl/blob/managed_services_version_with_sampling/tutorials/models/hetro-graph-with-node-features/ITSM_data_loader.py
.
   Note: The data loader uses a managed service version of ArangoDB. No
   installation of ArangoDB is necessary to run the examples. It is possible
   to load the entire data into the database, however, in the interest of
   making it possible to run the example quickly, we have loaded a sample of
   the data into the database.
   4. Model development: This is the contribution of this work. A
   heterogeneous graph convolution network with node features is implemented.
   All features were categorical. The discretization of numerical attributes
   is performed as part of pre-processing. An embedding layer is created to
   capture the embedding for each feature. Each node state is the combined
   representation of all its features. Therefore the input to each node is of
   the size = number of features for the node * size of the embedding for each
   feature.  The network consists of two layers. A cross-entropy loss was used
   for this problem.  20% of the data was used for testing and 80% of the data
   was used for training. The notebook
   https://github.com/arangoml/dgl/blob/managed_services_version_with_sampling/tutorials/models/hetro-graph-with-node-features/ITSM_Graph_Embedding.ipynb
contains
   the details of model development.

As a result of developing this model, the following outcomes are achieved:

   - We have a supervised learning model for prediction of the label of
   interest  - is an incident going to be reassigned in the course of its
   resolution?
   - We have embeddings for the nodes and the features associated with the
   nodes. These may be useful in their own right. For example, they can be
   used to determine how similar two incidents are using a similarity metric,
   for example, the euclidean distance. These embeddings can be captured and
   used for other machine learning applications using the same data. For
   example, clustering the support organization can yield insights into how
   the work in the support organization is segmented.


As mentioned earlier, this submission was intended but slightly early. I
thought I was submitting this to another repository. There will be more
documentation around these steps and perhaps some illustrations for the
results The above is a sketch of the story so far. If anything is unclear
or could use more information, please let me know. I have added Joerg, our
ML head to this thread.

Thanks for your response. We have really liked DGL!.

Rajiv

**Comment by rajivsam** [comment]
Hi Minjie,
Thanks. I have posted an explanation. As mentioned, I thought I was
submitting this to an internal repository. However, we did intend to
contribute to the examples. Basically we wanted a model for Heterogeneous
RGCN that supported node features. We only found an example for nodes
without features. So we explored creating one.  The problem and the
proposed solution are provided in brief. As mentioned, we do intend to add
more documentation to all steps. We would appreciate and welcome feedback.
Thank you
Rajiv

**Comment by rajivsam** [comment]
@jermainewang @BarclayII I have removed the data files and have created a README file with a description of the model. If you have any questions or need information or documentation on any aspect of the submission, please let me know.
Thanks 
Rajiv

**Comment by jermainewang** [comment]
HI @rajivsam , great effort! Sorry for not replying. I will spend time reading your codes and leave my thoughts.

**Comment by rajivsam** [comment]
> HI @rajivsam , great effort! Sorry for not replying. I will spend time reading your codes and leave my thoughts.

Hi @jermainewang . Thanks, look forward to it.

**Comment by rajivsam** [comment]
@jermainewang, I updated the PR with a change that makes the code simpler. We created a networkx adapter project for going between Arango Graphs and NetworkxGraphs. So to get the data for the graph, you just instantiate the networkx adapter and configure the adapter with information such as the db connection and query to get the data. The earlier commit had code to massage data from ArangoDB to Networkx. This is now abstracted away from this example.
